### PR TITLE
Fix for RA2 Mod: cloning vats

### LINKED ISSUE #17396: ClonesProducedUnits does not define a ProductionType.
It should override the ProductionType of the clone, otherwise there is no way to prevent the cloning structure be used as a regular production facility with `ClassicProductionQueues`, risking the loss of the clone.

### PR DESCRIPTION
**Pull request by torleif**
Fixes #17396

This fix will allow cloning vats in the RA2 mod to work correctly. 

The RA2 mod currently has a bug where the cloning vats don't work. This is because they're set to have a production type of Dummy, so that they can produce units. However there is a check in the OpenRA engine prevents new units of only the same type being produced. Because cloning vats should not be able to allow the construction of new units (you need a barracks to do this), it's set to a fake Dummy type. 

I was tempted to rip out the check altogether, however I'm unsure if other mods have buildings that use the cloning feature.

I've pulled the check out of the main function into its own function for readability.

### CODE REVIEW

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 1]
```diff
@@ -32,11 +32,32 @@ public class ClonesProducedUnits : ConditionalTrait<ClonesProducedUnitsInfo>, IN
...
+		   : base(info)
 		{
 			productionTraits = init.Self.TraitsImplementing<Production>().ToArray();
 		}
 
+		// will the trait allow me to clone a unit?
+		private bool ValidProductionTrait(string[] productionlist, string productionType)
+		{
+			if (string.IsNullOrEmpty(productionType))
+			{
```
Comment (tovl) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: Style nit: We prefer not to use brackets for single lines.

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 2]
```diff
@@ -32,11 +32,32 @@ public class ClonesProducedUnits : ConditionalTrait<ClonesProducedUnitsInfo>, IN
...
+				return true;
+			}
+
+			if (productionlist.Contains(productionType))
+			{
```
Comment (tovl) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: same here

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 3]
```diff
@@ -32,11 +32,32 @@ public class ClonesProducedUnits : ConditionalTrait<ClonesProducedUnitsInfo>, IN
...
+				return true;
+			}
+
+			if (productionlist.Contains("Dummy"))
+			{
```
Comment (tovl) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: and here

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 4]
```diff
@@ -55,8 +76,10 @@ public void UnitProducedByOther(Actor self, Actor producer, Actor produced, stri
 			// Stop as soon as one production trait successfully produced
 			foreach (var p in productionTraits)
 			{
-				if (!string.IsNullOrEmpty(productionType) && !p.Info.Produces.Contains(productionType))
+				if (!ValidProductionTrait(p.Info.Produces, productionType))
+				{
```
Comment (tovl) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: and here

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 5]
```diff
@@ -32,11 +32,32 @@ public class ClonesProducedUnits : ConditionalTrait<ClonesProducedUnitsInfo>, IN
...
 
 		public ClonesProducedUnits(ActorInitializer init, ClonesProducedUnitsInfo info)
-			: base(info)
+		   : base(info)
 		{
 			productionTraits = init.Self.TraitsImplementing<Production>().ToArray();
 		}
 
+		// will the trait allow me to clone a unit?
+		private bool ValidProductionTrait(string[] productionlist, string productionType)
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: Another style nit: `private` should be omitted, as it is the default.

**Comment by GraionDilach** [comment]
This is entirely wrong on principle.

The actual issue why the Vats break is #17396 (which this doesn't fix currently) and should be considered as a few-line fix - you need to add a `ProductionType` string with `FieldLoader.Require` property and then change the last `Produce` call within the `UnitProducedByOther`function to use this new `ProductionType` instead of the original.

See https://github.com/AttacqueSuperior/Engine/blob/master/OpenRA.Mods.AS/Duplicates/Traits/Buildings/ClonesProducedUnitsAS.cs as reference.

**Comment by Mailaender** [comment]
The code should probably move to https://github.com/openRA/ra2.

**Comment by torleif** [comment]
Thanks for your feedback. @GraionDilach I've rewritten the patch so it uses ProductionType instead. 

I've tested it with the following yml and it works as expected:

```
	Production:
		Produces: Clone
	ClonesProducedUnits:
		CloneableTypes: infantry
		ProductionType: Clone
```

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 6]
```diff
@@ -32,7 +36,7 @@ public class ClonesProducedUnits : ConditionalTrait<ClonesProducedUnitsInfo>, IN
 		readonly Production[] productionTraits;
 
 		public ClonesProducedUnits(ActorInitializer init, ClonesProducedUnitsInfo info)
-			: base(info)
+		   : base(info)
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: Please revert this change.

**Review by abcdefg30 — COMMENTED** [review]
Although untested, the code changes look good to me. Can you squash the commits, please?

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs (hunk 6 above)
Comment (torleif) [reply OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs]: thanks @abcdefg30, i've reverted the random extra tab removal

**Comment by torleif** [comment]
squashed commits

**Review by abcdefg30 — APPROVED** [review]

**Comment by GraionDilach** [comment]
Please edit the initial post to include the text "Fixes #17396" for ticket autoclosure.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b8ef66a35c61fbc169e820eded99ab9b45d7ee64...2a30b49122e5ae53a6da44e9fa7281904f8e60f1)
